[ur] feat(evm-utils/evm-height-rangemap):  evm-height-rangemap tool

dj8yfo opened the pull request:
#### Problem

#### Summary of Changes
1. send heights via channel
2. fetch height from cache or wait until fetched indefinetely (Option<H256>)
3. timeout in LedgerStorage
4. correct Err handling (panic in worker when giving up on error to fetch a new height, channel closed, panic in worker sending heights to channel)
5. BlockNotFound(..) results in None : Option<H256>
6. addded backon for retries with exponential backoff
7. split ranges into subranges for multiple LedgerStorage workers
8. add checking of nodes presence in rocksdb
9. add persisting of results to range_map.json of destination
Fixes #

Commit:
feat(evm-utils/evm-height-rangemap): evm-height-rangemap tool
1. send heights via channel
2. fetch height from cache or wait until fetched indefinetely (Option<H256>)
3. timeout in LedgerStorage
4. correct Err handling (panic in worker when giving up on error to fetch a new height, channel closed, panic in worker sending heights to channel)
5. BlockNotFound(..) results in None : Option<H256>
6. addded backon for retries with exponential backoff
7. split ranges into subranges for multiple LedgerStorage workers
8. add checking of nodes presence in rocksdb
9. add persisting of results to range_map.json of destination

vldm commented on evm-utils/evm-height-rangemap/src/storage.rs:
```diff
@@ -0,0 +1,41 @@
+    Primary(Primary),
+    Secondary(StorageSecondary),
+}
+impl Storage {
```
why we need special impl block for this tools.

Can we make a trait that will contain read-only operation, and reuse it?

vldm commented on evm-utils/evm-height-rangemap/src/range_map.rs:
```diff
@@ -0,0 +1,39 @@
+#[derive(Debug, Clone)]
+pub struct MasterRange {
+    file_path: PathBuf,
+    inner: Arc<Mutex<RangeMap<BlockNum, String>>>,
```
Do you really need Mutex there?

vldm commented on evm-utils/evm-height-rangemap/src/cached_height_index.rs:
```diff
@@ -0,0 +1,143 @@
+const MAX_TIMES: usize = 22; // 1 millisecond
+
+
+impl CachedHeightIndexBackgroundMan {
```
Didn't get the idea behind mooving this part to separate thread.

vldm reviewed (commented):
There a lot of problem in your current aproach:
1. Using vector for all range will store it in memory, which should be avoided.
Instead you can use reference implementation from rayon https://docs.rs/rayon/latest/rayon/iter/fn.split.html  
2. Now i see that MasterRange responsibility is to Load\Persist Range, so you can keep Mutex outside and make it single responsible.
3. You already use async code, but MasterRange is synchronized by Mutex
I think, that you can avoid Mutex, by implementing merge method in MapRange index.
Also i propose you to use rayon as basic fork-join parallelism library, or implement it in async as following:
```
let ranges_results = vec![];
for range in ranges {
ranges_results = spawn(async{some range work});
}
reduce/merge(range_results)

```

Commit:
chore(evm-block-recovery): add explicit error message

Commit:
chore(evm-block-recovery): use single connection for bigtable_fetcha
clone single connection for workers instead of using multiple distinct connections

Commit:
chore(evm-state::storage): create two_modes enum

vldm commented on evm-utils/evm-block-recovery/src/routines/scan_evm_state_roots.rs:
```diff
@@ -0,0 +1,82 @@
+use crate::{cli::ScanEvmStateRootsArgs, error::AppError};
+
+#[async_trait]
+pub trait BigtableFetcher {
```
Why we need Trait

vldm commented on evm-utils/evm-block-recovery/src/routines/scan_evm_state_roots.rs:
```diff
@@ -0,0 +1,82 @@
+        .await?;
+
+    let mut actual_len = 0;
+    while let Some((height, key)) = fetcha.get_block().await {
```
Move this logic into worker.

vldm replied:
And merge map at the end

vldm commented on evm-utils/evm-block-recovery/src/routines/scan_evm_state_roots.rs:
```diff
@@ -0,0 +1,82 @@
+            if present {
+                Some("X")
+            } else {
+                None
```
Let's name values "Present"/"No root"/"No block"

vldm commented on evm-utils/evm-block-recovery/src/routines/scan_evm_state_roots.rs:
```diff
@@ -0,0 +1,82 @@
+        };
+        if let Some(result) = result {
+            rangemap.update(height, result.to_string())?;
+            log::debug!("{} -> {:?}, {}", height, key, result,);
```
trace

vldm commented on evm-utils/evm-block-recovery/src/routines/scan_evm_state_roots/bigtable_fetcha.rs:
```diff
@@ -0,0 +1,157 @@
+            *lock += 1;
+        }
+        async {
+            log::debug!(
```
trace

vldm commented on evm-utils/evm-block-recovery/src/routines/scan_evm_state_roots/range_map.rs:
```diff
@@ -0,0 +1,38 @@
+        drop(inner);
+        Ok(())
+    }
+}
```
new line

vldm commented on evm-utils/evm-block-recovery/src/routines/scan_evm_state_roots/range_map.rs:
```diff
@@ -0,0 +1,38 @@
+#[derive(Debug, Clone)]
+pub struct MasterRange {
+    file_path: PathBuf,
+    inner: Arc<Mutex<RangeMap<BlockNum, String>>>,
```
No mutex

Commit:
chore(evm-block-recovery): routines::scan_evm_state_roots review fixes 2
